keras_dqn/agent.py

```diff
@@ -0,0 +1,92 @@
+import numpy as np
+from keras.utils.np_utils import to_categorical
+
+
+class Agent(object):
+    def __init__(self, model, target_model=None, update_model_every=1,
+                 update_target_every=1, min_n_frames=1, frames_per_epoch=1,
+                 eps=.1, eps_rate=1, min_eps=0, test_eps=0, rng_seed=123):
+        """
+        Parameters:
+        -----------
+        model: keras model (agent)
+        target_model: keras model (desired)
+        update_target_every : number of frames skipped before syncing the target with model
+        update_model_every" : number of frames skipped before gradient step
+        min_n_frames: number of frames stored to memory before start training
+        frames_per_epoch : how long is an epoch?
+        eps: probability of random action (epsilon greedy)
+        eps_rate: epsilon annealing rate
+        min_eps: min epsilon
+        test_eps: probability of random action during test
+        rng_seed
+        """
+        self.rng_seed = rng_seed
+        np.random.seed(rng_seed)
+
+        self.model = model
+        if target_model is None:
+            self.target_model = model
+        else:
+            self.target_model = target_model
+            self.target_swap = True  # external target flag
+            print("Using target model update.")
+        self.update_model_every = update_model_every
+        self.update_target_every = update_target_every
+
+        # memory info
+        self.min_n_frames = min_n_frames
+
+        # epsilon greedy info
+        self.eps = eps
+        self.eps_rate = eps_rate
+        self.min_eps = min_eps
+        self.test_eps = test_eps
+
+    @property
+    def n_actions(self):
+        return self.model.output_shape[-1]
+
+    @property
+    def input_shape(self):
+        return self.model.input_shape
+
+    def get_action(self, observation, train=False):
+        rnd = np.random.rand()
+        rnd_action = np.random.randint(0, self.n_actions)
+        if train and rnd < self.eps:
+            return rnd_action
+        elif train is False and rnd < self.test_eps:
+            return rnd_action
+        else:
+            # this code is meant to work only if doing a single action!
+            q = self.model.predict(observation)
+            # random tie breaker
+            max_inds = np.argwhere(q == np.max(q))
+            return np.random.choice(max_inds.reshape(-1,))
+
+    def train_on_batch(self, state, action, reward, next_state, terminal, gamma=.99, accum_mode="mean"):
```

Comment:
Will we have the agent create and handle the memory internally or externally? If its internally we should need to pass state, action, etc in, or we can and just make this a private function?